webrtc/go/main.go

```diff
@@ -0,0 +1,312 @@
+	//   * a control network
+	//
+	// The data network is where standard test traffic flows. The control
+	// network connects us ONLY with the sync service, InfluxDB, etc. All
+	// traffic shawebrtc rules are applied to the data network. Thanks to this
+	// separation, we can simulate disconnected scenarios by detaching the data
+	// network adapter, or blocking all incoming/outgoing traffic on that
+	// network.
+	//
+	// We need to listen on (and advertise) our data network IP address, so we
+	// obtain it from the NetClient.
+	ip := initCtx.NetClient.MustGetDataNetworkIP()
+
+	// ☎️  Let's construct the libp2p node.
+	listenAddr := fmt.Sprintf("/ip4/%s/udp/0/webrtc", ip)
+	host, err := NewLibp2p(ctx,
+		secureChannel,
+		libp2p.Transport(webrtc.New),
+		libp2p.ListenAddrStrings(listenAddr),
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to instantiate libp2p instance: %w", err)
+	}
+	defer host.Close()
+
+	// 🚧  Now we instantiate the webrtc service.
+	//
+	// This adds a stream handler to our Host so it can process inbound pings,
+	// and the returned WebrtcService instance allows us to perform outbound pings.
+	ping := ping.NewPingService(host)
+
+	// Record our listen addrs.
+	runenv.RecordMessage("my listen addrs: %v", host.Addrs())
+
+	// Obtain our own address info, and use the sync service to publish it to a
+	// 'peersTopic' topic, where others will read from.
+	var (
+		hostId = host.ID()
+		ai     = &PeerAddrInfo{ID: hostId, Addrs: host.Addrs()}
+
+		// the peers topic where all instances will advertise their AddrInfo.
+		peersTopic = sync.NewTopic("peers", new(PeerAddrInfo))
+
+		// initialize a slice to store the AddrInfos of all other peers in the run.
+		peers = make([]*PeerAddrInfo, 0, runenv.TestInstanceCount)
+	)
+
+	// Publish our own.
+	initCtx.SyncClient.MustPublish(ctx, peersTopic, ai)
+
+	// Now subscribe to the peers topic and consume all addresses, storing them
+	// in the peers slice.
+	peersCh := make(chan *PeerAddrInfo)
+	sctx, scancel := context.WithCancel(ctx)
+	sub := initCtx.SyncClient.MustSubscribe(sctx, peersTopic, peersCh)
+
+	// Receive the expected number of AddrInfos.
+	for len(peers) < cap(peers) {
+		select {
+		case ai := <-peersCh:
+			peers = append(peers, ai)
+		case err := <-sub.Done():
+			return err
+		}
+	}
+	scancel() // cancels the Subscription.
+
+	// ✨
+	// ✨  Now we know about all other libp2p hosts in this test.
+	// ✨
+
+	// This is a closure that webrtcs all peers in the test in parallel, and
+	// records the latency value as a message and as a result datapoint.
+	webrtcPeers := func(tag string) error {
+		g, gctx := errgroup.WithContext(ctx)
+		for _, ai := range peers {
+			if ai.ID == hostId {
+				continue
+			}
+
+			id := ai.ID // capture the ID locally for safe use within the closure.
+
+			g.Go(func() error {
+				// a context for the continuous stream of webrtcs.
+				pctx, cancel := context.WithCancel(gctx)
+				defer cancel()
+				res := <- ping.Ping(pctx, id)
```

Comment:
This is a naked channel block. You do not want to pair it with something that can time out or w/e?

Reply:
If it gets stuck here, the test itself will eventually time out (by order of testground daemon), so I'm not terribly worried about it. But if you think it's important I'm open to the idea.

Reply:
Dunno, might be more useful to have a decent helpful error that you can explicitly throw, than some vague testdaemon timeout